src/Client/ConnectionString.cpp

```diff
@@ -0,0 +1,240 @@
+        // several users: 'usr1@host1,@usr2@host2' is invalid.
+        if (hosts_end_pos != std::string_view::npos && connection_string[hosts_end_pos] == '@')
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                            "Symbols '@' in URI in password or user name should be percent-encoded. Individual user names for different hosts also prohibited. {}",
+                            connection_string);
+        }
+    }
+    else
+        hosts_end_pos = hosts_or_user_info_end_pos;
+
+    const auto * hosts_end = hosts_end_pos != std::string_view::npos ? connection_string.begin() + hosts_end_pos
+                                                                     : connection_string.end();
+
+    try
+    {
+        /** Poco::URI doesn't support several hosts in URI.
+          * Split string clickhouse:[user[:password]@]host1:port1, ... , hostN:portN[database]?[query_parameters]
+          * into multiple string for each host:
+          * clickhouse:[user[:password]@]host1:port1[database]?[query_parameters]
+          * ...
+          * clickhouse:[user[:password]@]hostN:portN[database]?[query_parameters]
+          */
+        Poco::URI uri;
+        const auto * last_host_begin = connection_string.begin() + offset;
+        for (const auto * it = last_host_begin; it != hosts_end; ++it)
+        {
+            if (*it == ',')
+            {
+                buildConnectionString(uri, hosts_and_ports_arguments, {last_host_begin, it}, {hosts_end, connection_string.end()});
+                last_host_begin = it + 1;
+            }
+        }
+
+        if (uri.empty())
+        {
+            // URI has no host specified
+            uri = std::string(connection_string);
+            getHostAndPort(uri, hosts_and_ports_arguments);
+        }
+        else
+            buildConnectionString(uri, hosts_and_ports_arguments, {last_host_begin, hosts_end}, {hosts_end, connection_string.end()});
+
+        Poco::URI::QueryParameters params = uri.getQueryParameters();
+        for (const auto & param : params)
+        {
+            if (param.first == "secure" || param.first == "s")
+            {
+                if (!param.second.empty())
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "secure URI query parameter does not require value");
+
+                common_arguments.push_back(makeArgument(param.first));
+            }
+            else
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "URI query parameter {} is unknown", param.first);
+        }
+
+        auto user_info = uri.getUserInfo();
+        if (!user_info.empty())
+        {
+            // Poco::URI doesn't decode user name/password by default.
+            // But ClickHouse allows to have users with email user name like: 'john@some_mail.com'
+            // john@some_mail.com should be percent-encoded: 'john%40some_mail.com'
+            std::string::size_type pos = user_info.find(':');
```

Comment:
Not wrong but let's simplify to type `size_t`.

Reply:
Done